### ISSUE TITLE
Releng for Granite

### ISSUE THREAD

**Issue by liuche** [mozilla-mobile/focus-android#1459]
We want to do a rotating releng task amongst the team members.

This issue is to go through the checklist: https://github.com/mozilla-mobile/focus-android/wiki/Sprint-checklist

**Comment by Sdaswani** [comment]
Reproducing the checklist here for @liuche to track:

See [Focus for Android Train Schedule](https://wiki.mozilla.org/Mobile/Focus/Android/Train_Schedule)

## Sprint start

- [ ] Update versionName in build.gradle for the next milestone (preliminary version number)

## String freeze

- [ ] Make sure all issues with label "strings needed" have strings assigned and the label updated to "strings approved"
- [ ] Pre-land all strings for features that have not been implemented yet
- [ ] Export strings and open PR in the L10N repository

## Feature complete

- [ ] Create a branch for the current milestone. After that master is tracking the next milestone. Usually done on the Monday after feature complete on Friday.

## L10N sign-off

- [ ] Import latest strings from L10N repository (into the milestone branch)
- [ ] Remove locales that should not ship from the milestone branch

## Submission

- [ ] Tag first RC version in Github. For 1.0 the tag would be v1.0-RC1
- [ ] Build release versions of Klar and Focus
- [ ] Create Bugzilla bug for signing and uploading the builds
- [ ] Let release management upload the builds to the Alpha channel (internal distribution)
- [ ] If the builds passed a smoke test let release management promote the build to the beta channel (public distribution)
- [ ] Upload the APK to the [releases page](https://github.com/mozilla-mobile/focus-android/releases).

## During Beta

- [ ] Check Google Play for new crashes. File issues and triage.
- [ ] If bugs are considered release blocker then fix them on master and the milestone branch (cherry-pick / uplift)
- [ ] If needed tag a new RC version (e.g. v1.0-RC2) and follow the submission checklist again.

## Release

- [ ] Release management promotes the build from beta to the release channel. For minor releases this can happen at any time during the day. Major releases often need to be synchronized with other marketing activities (e.g. blog postings). Releases are rolled out to 99% of the population (otherwise the rollout can't be stopped).

## After the release

- [ ] After some days release management sets the roll out to 100%
- [ ] Check whether there are new crashes that need to be filed.
- [ ] Tag the last and released RC version additionally with the tag of the release (v1.0-RC2 -> v1.0)
- [ ] Upload signed APKs to the [release page](https://github.com/mozilla-mobile/focus-android/releases)

**Comment by pocmo** [comment]
> Update versionName in build.gradle for the next milestone (preliminary version number)

I checked this one. I did that yesterday already.

**Comment by liuche** [comment]
@pocmo I have some questions for you :)

Can you cc me on the Bugzilla bug for signing and uploading the Folly RC build, so I can see what format/component the bug should be in? (Have the alpha/beta already been uploaded? I didn't see an RC tag on Github for the Folly branch or anything on the Play Console, so I assumed not.)

Also, when trying out building the release builds (for the step to upload them to bugzilla to be signed), do you do that through Android Studio? AS complains that I need to have a signing configuration before building - what signing configuration do you use?

`Error running app: The apk for your currently selected variant (app-focus-webview-release-unsigned.apk) is not signed. Please specify a signing configuration for this variant (focus-webview-release)`

Finally, for the step about looking at the crashes, do you just use the Play Console, and look at the Crashes section from the main dashboard? Are there other places that you check?

**Comment by pocmo** [comment]
Great questions! :)


> Can you cc me on the Bugzilla bug for signing and uploading the Folly RC build, so I can see what format/component the bug should be in?

Absolutely! I was going to add some of them to the docs. Here are the last ones:
* 2.1 RC1: https://bugzilla.mozilla.org/show_bug.cgi?id=1403930
* 2.0 RC2: https://bugzilla.mozilla.org/show_bug.cgi?id=1399831
* 2.0 RC1: https://bugzilla.mozilla.org/show_bug.cgi?id=1398055

> (Have the alpha/beta already been uploaded? I didn't see an RC tag on Github for the Folly branch or anything on the Play Console, so I assumed not.)

No, we upload our first release build (RC 1) to the beta channel on submission day:
https://wiki.mozilla.org/Mobile/Focus/Android/Train_Schedule

For Folly this would be October 13th. The day before is the L10N sign-off - so technically we can't create a release build yet because the strings can still change.

However as your only responsibility is *Granite* here I can continue pushing Folly to release and update the docs. There is always some overlap: One version is made ready for release and the next one is already in development. I'll file a separate bug for Folly to make things more clear. :)

> Also, when trying out building the release builds (for the step to upload them to bugzilla to be signed), do you do that through Android Studio? AS complains that I need to have a signing configuration before building - what signing configuration do you use?

I still do that on the command line like I described in the mail for the first release (when I was on PTO). For two reasons:
* I have to set the environment variable for the Adjust token. Technically I could do that in Android Studio; but for me it's more obvious what happens on the command line.
* I can create two unsigned release builds easily with just a command.

I'll forward the old mail to you (the token is in there too). But in a nutshell I do:
```bash
git log // git status // git diff
# ^- Make sure I'm on the right branch and there are not local changes
#     It's easy to accidentally compile local changes into the release... :)

export ADJUST_TOKEN_FOCUS="..."
echo $ADJUST_TOKEN_FOCUS
# ^-- I'm paranoid so I'll ways check that it's set correctly

./gradlew clean
# ^-- Making sure that there are no weird caches hanging around.
#       In addition to that the output will show whether gradle found the token.

./gradlew assembleRelease
# After that the APKs will be in app/build/outputs/apk

# To be really sure I didn't fuck up I *copy* the APKs and sign the copy with my local debug key (Android Studio created this one).
# I then install them and do a quick smoke test on my device. We deliberately crash focus release builds if they do not contain an
# Adjust token - so that we avoid releasing them without a token accidentally. So this test also makes sure that I didn't create
# a crashing build.. :)
# e.g.:
jarsigner -verbose -keystore ~/.android/debug.keystore -storepass android -keypass android app-focus-webview-release-unsigned.apk androiddebugkey
adb install -r app-focus-webview-release-unsigned.apk
# (Make sure there's no build with the release signature installed)



> Finally, for the step about looking at the crashes, do you just use the Play Console, and look at the Crashes section from the main dashboard? Are there other places that you check?

For beta builds I look at the crashes for this specific version (Android vitals -> ANR & Crashes -> Crashes tab -> Version drop down). Especially crashes with "New in version XXXXX" are interesting.

Sometimes it can be helpful to look at the number of Beta testers too (To get a rough feeling about how serious the crash is):
Release management ->  App releases -> Manage Beta -> (Green number, top right).

As soon as the version is released to 99% I go to the release dashboard and compare the new version with the previous version to see how/if our crash rate changes and what users comment. And then later at the crashes again - just because there are more users on the release version and some crashes might show up more now.

**Comment by pocmo** [comment]
Folly: #1464

**Comment by Sdaswani** [comment]
Thanks for the good questions and extra detail @liuche and @pocmo . I think it makes sense for the two RM hats to co-exist during the 'QA week' that precedes release, and for the previous RM to own that until completion while the new RM starts on the next milestone.

**Comment by pocmo** [comment]
You are quite early with the L10N import. October 26th is L10N sign-off; theoretically the translators still have some days. But of course you can do multiple imports if there are still changes. :)

**Comment by liuche** [comment]
Oh! Good call - it wasn't obvious to me what the dates were actually, and I remember Delphine saying the translators like seeing the screenshots to verify their translations. I'll be sure to reimport again before tagging.

Also, @pocmo, just to double-check, I won't need to uplift the taskcluster patches to auto-build release apks, because the Taskcluster integration is with master, right?

**Comment by pocmo** [comment]
> Also, @pocmo, just to double-check, I won't need to uplift the taskcluster patches to auto-build release apks, because the Taskcluster integration is with master, right?

That's a very good question. I just checked and it looks like taskcluster will not be triggered if .taskcluster.yml on this branch does not contain a release configuration. I can create a PR to uplift those patches. As they do not touch app code the risk should be pretty low. I'll flag you for "review" and then you can decide as release manager. :)

**Comment by liuche** [comment]
@pocmo you also mentioned that there are some new locales that have reached 100% - where do you look for that? Also, I don't see any instructions for adding new locales in the wiki. Do these steps make sense:
- Import new locale directories into the repo
- Update tools/l10n/locales.py

Are there any other steps? (These aren't Play Store specific locales, just in-app, so we don't need to do anything about Play Store/marketing, I assume.) I'll add these steps to the wiki for "Adding New Locales".

**Comment by Sdaswani** [comment]
Not sure about the steps but you can find the localization status [here](https://l10n.mozilla-community.org/webstatus/?product=focus-android).

**Comment by pocmo** [comment]
I look at Pontoon for the state of things:
https://pontoon.mozilla.org/projects/focus-for-android/

>    Import new locale directories into the repo
>    Update tools/l10n/locales.py

Yeah, that's correct. Our importer will import all translations. For releases we will filter the list based on the locales in locales.py. So adding new locales there (or moving them from the "screenshot-only" list) should be enough for them to get added to the release.

> Are there any other steps? (These aren't Play Store specific locales, just in-app, so we don't need to do anything about Play Store/marketing, I assume.) I'll add these steps to the wiki for "Adding New Locales".

Yeah, those are completely separate and not maintained by us. They live here:
https://l10n.mozilla-community.org/stores_l10n/

Sylvestre has a script to automatically update them all if needed.

**Comment by pocmo** [comment]
I just noticed that the translation state is now a bit harder to see from the dashboard(s). String Freeze and L10N sign-off are on the same day and back when I was the only one handling releases I would first import all the locales for the release and remove everything that is not needed for the release (what the script now does). And after that I would do the new export.

This time it happened the other way around and it makes things harder: When importing the final strings there are suddenly a lot of unused (new) strings and lint goes crazy. And in addition to that you can't see what languages are at 100% because new strings have just landed.

Maybe we should make this more explicit in the release checklist so that two release managers do not interfere with each other on that day?

**Comment by liuche** [comment]
> This time it happened the other way around and it makes things harder: When importing the final strings there are suddenly a lot of unused (new) strings and lint goes crazy. And in addition to that you can't see what languages are at 100% because new strings have just landed.

Hmmm, yeah I'm realizing that since export/string-freeze (Wednesday) now happens one day before l10n-signoff/import (Thursday), but the l10n repo doesn't freeze, there will definitely be new strings, unless we shift import to be before export - the only way to fix that is to make them the same day again, right?

There is an actual (kind of nitpicky) problem when there are string deletions/replacement (as happened with the 2.4 export) - a previously translated string will actually become unavailable for localizers that were quick and localized the new string during that one day!

In the future, for this specific case - string replacements - we should could do one of the following:
a) flag the accompanying code change to be uplifted (to not penalize quick localizers), and because ideally it's just a quick string change
b) (non-ideally) just ship with old string, and ship the untranslated string
c) land string changes "late" so they're exported in the same milestone (which is the same as a, but doesn't require extra releng work)
d) exclude the string replacement changes from the l10n import (this sounds manual)
e) have sign-off and export on the same day (I talked to delphine about this, will follow up later)

For this release, I actually had an old l10n import from Monday before the string changes, so I talked to delphine and we agreed to only import the new ko strings since Monday because all the new strings from last sprint are non-visible, and ko was the only locale that had significant changes since Monday.

**Comment by liuche** [comment]
Okay, the release build script is running! https://github.taskcluster.net/v1/repository/mozilla-mobile/focus-android/granite/latest

**Comment by Sdaswani** [comment]
Wow this seems complicated. :/

**Comment by liuche** [comment]
Oh, I see what you mean about the line errors :) I'll just fix those by hand.

**Comment by liuche** [comment]
Made a copy of the release build to sign and test, tried out the ko locale as well.
Bugzilla bug for signing: https://bugzilla.mozilla.org/show_bug.cgi?id=1412220

@pocmo I'm PTO tomorrow, but I think everything should be all sorted out :)

**Comment by pocmo** [comment]
>  e) have sign-off and export on the same day (I talked to delphine about this, will follow up later)

Yeah, the same day; or in other words: sign-off always before export (could also happen multiple days before; if this makes sense).

> Okay, the release build script is running! https://github.taskcluster.net/v1/repository/mozilla-mobile/focus-android/granite/latest

It's green! 💚 This makes me very happy. I wonder why it's RC3 though? We didn't build RC1 - RC2, or did we?

> @pocmo I'm PTO tomorrow, but I think everything should be all sorted out :)

Looks like it's signed and uploaded to the Alpha track already! 🙌 I'll do some smoke testing on my device and ask them to promote it to the Beta branch. :)

**Comment by pocmo** [comment]
> It's green! 💚 This makes me very happy. I wonder why it's RC3 though? We didn't build RC1 - RC2, or did we?

Oh, now I saw that there are actually multiple builds! :)

As long as we didn't distribute a build to any user yet I took the freedom to actually delete tags/releases again in case I ran into an error. Only if something actually went out to get signed and uploaded I made a new RC. Only once we had to deploy a new build to production (release channel, not beta) to fix an issue. In this case I also updated the version number (1.2 -> 1.2.1).

**Comment by pocmo** [comment]
Got the update from the Alpha channel. Couldn't find any issues in a quick smoke test. Flagged release management on Bugzilla to promote the build from the Alpha channel to the Beta channel.

**Comment by liuche** [comment]
Seeing a 2.3% crash rate on beta (which is higher than prod), the highest crashers are webview crashes, but mostly on older versions.

There are two new bugs: #1610 and #1611.

However, looking at the commits that actually landed in 2.3, I don't see that much user-facing code is in this release - most of the changes are build-side improvements, so I don't see any culprits for this crash rate. (Is the crash rate usually higher in Beta, @pocmo? Is this out of the ordinary?)

Since I don't see any clear regressions, or any causes for new crashers, I don't see a reason to hold Beta back. I assume that Releng releases Beta to Prod automatically, and this doesn't require any more action from us.

**Comment by pocmo** [comment]
> Seeing a 2.3% crash rate on beta (which is higher than prod), the highest crashers are webview crashes, but mostly on older versions.

I filed #1630 - There an API on newer versions to recover from webview crashes. On older versions we just have to live with them. :(

> There are two new bugs: #1610 and #1611, but will investigate this, and they only have a few crashes total.

Yeah, they seem to be super low - I wouldn't block the release for them.

> (Is the crash rate usually higher in Beta, @pocmo? Is this out of the ordinary?)

Unfortunately I couldn't see the crash rate for Beta builds previously. This looks like it's a new feature in Google Play? :)

> Since I don't see any clear regressions, or any causes for new crashers, I don't see a reason to hold Beta back. I assume that Releng releases Beta to Prod automatically, and this doesn't require any more action from us.

I usually ping Sylvestre in the release bug to let him know that we looked at the Beta crash rate etc. and are ready to release.

**Comment by pocmo** [comment]
I pinged Sylvestre on the Bugzilla bug and asked for a slow roll out to just 20% of the user population. After all there's no need to rush the dot release - let's see how the crash rate looks for a subset of the release population:
https://bugzilla.mozilla.org/show_bug.cgi?id=1412220

**Comment by liuche** [comment]
Currently rolled out to 99%, crash rate hasn't gone up (although there is more variance in the ANR rate).

**Comment by liuche** [comment]
Published to Focus releases page, which closes out the checklist.

**Comment by pocmo** [comment]
@liuche Back when we create this issue we didn't have bullet points for uploading the builds to S3 so that they show up on archive.mozilla.org.

I've been using the S3 command line tool for that:
https://aws.amazon.com/cli/

Did you get credentials for S3 via (encrypted) mail?

**Comment by liuche** [comment]
I'm waiting on keys from oremj for a new gpg key, because my gpg passcode didn't make it onto my keychain. Can you upload the 2.3 builds, @pocmo? I'll get it set up for the next time!

**Comment by pocmo** [comment]
Yeah, I can take this over :)

**Comment by pocmo** [comment]
Uploaded 2.3 builds to archive.mozilla.org
